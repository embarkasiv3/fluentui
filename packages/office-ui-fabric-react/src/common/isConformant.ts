import { isConformant as isConformantBase, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: IsConformantOptions) {
  const defaultOptions = {
    disabledTests: ['has-docblock'],
  };

  isConformantBase(defaultOptions, testInfo);
}
