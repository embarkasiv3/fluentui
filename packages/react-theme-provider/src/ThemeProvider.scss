/**
 * The ThemeProvider has a root consuming class which will apply the current
 * token values to the container DIV, so that fonts and default colors, which
 * refer to token values, can be applied.
 */
.root {
  background: var(--body-fill);
  color: var(--body-text);
  font-family: var(--body-fontFamily);
  font-weight: var(--body-fontWeight);
  font-size: var(--body-fontSize);
  line-height: var(--body-fontLineHeight);

  > * {
    box-sizing: border-box;
  }
}
