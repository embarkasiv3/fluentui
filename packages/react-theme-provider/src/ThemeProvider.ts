export * from './components/ThemeProvider/index';
