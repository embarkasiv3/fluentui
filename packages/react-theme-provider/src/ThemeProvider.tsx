import { useMergedRefs } from '@fluentui/react-hooks';
import { mergeThemes, themeToCSSVariables, PartialTheme, Theme } from '@fluentui/react-theme';
import { getSlots, makeMergeProps } from '@fluentui/react-utils';
import * as React from 'react';

import { internal__ThemeContext, useTheme } from './context';

export interface ThemeProviderProps extends React.HTMLAttributes<HTMLElement> {
  theme?: PartialTheme | Theme;
}
export interface ThemeProviderState extends React.HTMLAttributes<HTMLElement> {
  theme: Theme;
}

const mergeProps = makeMergeProps<ThemeProviderState>();

export function useThemeProviderState(draftState: ThemeProviderState) {
  const parentTheme = useTheme();
  const localTheme = draftState.theme;

  draftState.theme = mergeThemes(parentTheme, localTheme);
  draftState.style = React.useMemo(() => {
    // TODO: should we consider insertion to head?
    //       - how to modify, remove styles?
    //       - SSR rendering

    // TODO: what variables should be rendered? Merged or only changed?
    // TODO: how we will proceed with Portals?
    return {
      ...draftState.style,
      ...themeToCSSVariables(draftState.theme),
    };
  }, [draftState.style, draftState.theme]);
}

export function renderThemeProvider(state: ThemeProviderState) {
  const { slots, slotProps } = getSlots(state);
  const { theme } = state;

  return (
    <internal__ThemeContext.Provider value={theme}>
      <slots.root {...slotProps.root} />
    </internal__ThemeContext.Provider>
  );
}

/**
 * Returns the ThemeProvider render function and calculated state, given user input, ref, and
 * a set of default prop values.
 */
export function useThemeProvider(props: ThemeProviderProps, ref: React.Ref<HTMLElement>) {
  const rootRef = useMergedRefs(ref, React.useRef<HTMLElement>(null));
  const state = mergeProps(
    {
      ref: rootRef,
      as: 'div',
    },
    {},
    props,
  );

  useThemeProviderState(state);

  return {
    state,
    render: renderThemeProvider,
  };
}

/**
 * Used to provide CSS variables to DOM and theme tokens via React Context.
 */
export const ThemeProvider: React.FunctionComponent<ThemeProviderProps> = React.forwardRef<
  HTMLDivElement,
  ThemeProviderProps
>((props: ThemeProviderProps, ref: React.Ref<HTMLDivElement>) => {
  const { render, state } = useThemeProvider(props, ref);

  return render(state);
});

ThemeProvider.displayName = 'ThemeProvider';
