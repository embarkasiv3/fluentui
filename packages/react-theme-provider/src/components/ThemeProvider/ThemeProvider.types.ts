import * as React from 'react';
import { ComponentProps } from '@fluentui/react-utilities';
import { PartialTheme, Theme } from '@fluentui/react-theme';

export interface ThemeProviderProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  theme?: PartialTheme | Theme;
}

export interface ThemeProviderState extends ThemeProviderProps {
  /**
   * Ref to the root slot
   */
  ref: React.MutableRefObject<HTMLElement>;
  theme: Theme;
}
