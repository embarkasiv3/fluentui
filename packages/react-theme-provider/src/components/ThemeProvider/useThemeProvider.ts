import * as React from 'react';
import { makeMergePropsCompat, resolveShorthandProps, useMergedRefs } from '@fluentui/react-utilities';
import { ThemeProviderProps, ThemeProviderState } from './ThemeProvider.types';
import { useTheme } from '@fluentui/react-shared-contexts';
import { mergeThemes, themeToCSSVariables } from '@fluentui/react-theme';

export const themeProviderShorthandProps: (keyof ThemeProviderProps)[] = [];

// eslint-disable-next-line deprecation/deprecation
const mergeProps = makeMergePropsCompat<ThemeProviderState>({ deepMerge: themeProviderShorthandProps });

/**
 * Create the state required to render ThemeProvider.
 *
 * The returned state can be modified with hooks such as useThemeProviderStyles,
 * before being passed to renderThemeProvider.
 *
 * @param props - props from this instance of ThemeProvider
 * @param ref - reference to root HTMLElement of ThemeProvider
 * @param defaultProps - (optional) default prop values provided by the implementing type
 */
export const useThemeProvider = (
  props: ThemeProviderProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: ThemeProviderProps,
): ThemeProviderState => {
  const state = mergeProps(
    {
      ref: useMergedRefs(ref, React.useRef(null)),
      as: 'div',
    },
    defaultProps,
    resolveShorthandProps(props, themeProviderShorthandProps),
  );

  const parentTheme = useTheme();
  const localTheme = state.theme;

  state.theme = mergeThemes(parentTheme, localTheme);
  state.style = React.useMemo(() => {
    // TODO: should we consider insertion to head?
    //       - how to modify, remove styles?
    //       - SSR rendering

    // TODO: what variables should be rendered? Merged or only changed?
    // TODO: how we will proceed with Portals?
    return {
      ...state.style,
      ...themeToCSSVariables(state.theme),
    };
  }, [state.style, state.theme]);

  return state;
};
