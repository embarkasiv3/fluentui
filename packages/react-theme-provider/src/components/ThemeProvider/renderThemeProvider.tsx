import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { ThemeProviderState } from './ThemeProvider.types';
import { themeProviderShorthandProps } from './useThemeProvider';
import { ThemeContext } from '@fluentui/react-shared-contexts';

/**
 * Render the final JSX of ThemeProvider
 */
export const renderThemeProvider = (state: ThemeProviderState) => {
  const { slots, slotProps } = getSlots(state, themeProviderShorthandProps);

  return (
    <ThemeContext.Provider value={state.theme}>
      <slots.root {...slotProps.root}>{state.children}</slots.root>
    </ThemeContext.Provider>
  );
};
