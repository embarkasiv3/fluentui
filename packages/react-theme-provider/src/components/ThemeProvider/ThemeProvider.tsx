import * as React from 'react';
import { useThemeProvider } from './useThemeProvider';
import { ThemeProviderProps } from './ThemeProvider.types';
import { renderThemeProvider } from './renderThemeProvider';

export const ThemeProvider = React.forwardRef<HTMLElement, ThemeProviderProps>((props, ref) => {
  const state = useThemeProvider(props, ref);
  return renderThemeProvider(state);
});

ThemeProvider.displayName = 'ThemeProvider';
