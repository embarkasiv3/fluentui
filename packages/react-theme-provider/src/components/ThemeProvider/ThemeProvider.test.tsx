import * as React from 'react';
import { ThemeProvider } from './ThemeProvider';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';
import { ThemeContext } from '@fluentui/react-shared-contexts';

describe('ThemeProvider', () => {
  isConformant({
    Component: ThemeProvider,
    displayName: 'ThemeProvider',
    helperComponents: [ThemeContext.Provider],
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for ThemeProvider in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<ThemeProvider>Default ThemeProvider</ThemeProvider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
