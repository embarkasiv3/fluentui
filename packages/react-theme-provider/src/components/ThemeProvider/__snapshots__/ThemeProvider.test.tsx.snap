// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider renders a default state 1`] = `
<div
  style={Object {}}
>
  Default ThemeProvider
</div>
`;
